### PR TITLE
Feature/69 add s3 service interface

### LINKED ISSUE #69: Add S3 Service interface

### PR DESCRIPTION
**Pull request by Reza-Noei**
Please review this PR 🌹

fixes #69

### CODE REVIEW

File: Source/BSN.Commons.S3.Abstraction/BSN.Commons.S3.Abstraction.csproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (soroshsabz) [inline Source/BSN.Commons.S3.Abstraction/BSN.Commons.S3.Abstraction.csproj]: I think maybe is better to merge this project into BSN.Commons

Comment (soroshsabz) [reply Source/BSN.Commons.S3.Abstraction/BSN.Commons.S3.Abstraction.csproj]: what is your opinion?

File: Source/BSN.Commons.S3.Abstraction/BucketPathBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Text;
+
+namespace BSN.Commons.S3.Abstraction
+{
+    public class BucketPathBuilder
+    {
+        public static string Build(params string[] buckets)
```
Comment (soroshsabz) [inline Source/BSN.Commons.S3.Abstraction/BucketPathBuilder.cs]: Please add some comments to describe schema of path

File: Source/BSN.Commons.S3.Abstraction/BulkCredentialModel.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace BSN.Commons.S3.Abstraction
+{
+    [Obsolete("[R.Noei] I'm not sure if we need this anymore.")]
+    public class BulkCredentialModel
+    {
+        public BulkCredentialModel(string endPoint,
```
Comment (soroshsabz) [inline Source/BSN.Commons.S3.Abstraction/BulkCredentialModel.cs]: Add constructor comments, and use inhertidoc for param (such as SAF)

File: Source/BSN.Commons.S3.Abstraction/S3ServiceOptions.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿namespace BSN.Commons.S3.Abstraction
+{
+    public class S3ServiceOptions
```
Comment (soroshsabz) [inline Source/BSN.Commons.S3.Abstraction/S3ServiceOptions.cs]: Please add comments based on S3 site docs (maybe you can link them too)

File: Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{5F798AA1-1A65-4C9F-8415-3100C141DD87}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>BSN.Commons.S3.Net</RootNamespace>
+    <AssemblyName>BSN.Commons.S3.Net</AssemblyName>
+    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
```
Comment (soroshsabz) [inline Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj]: Why we need two project for .Netframework and Core?

As I know, S3 support both of them, so why do not create one project with .NET Standard 2.0 for S3 implementation?

Comment (soroshsabz) [reply Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj]: https://www.nuget.org/packages/AWSSDK.S3

**Review by soroshsabz — CHANGES_REQUESTED** [review]

File: Source/BSN.Commons.S3.Abstraction/BSN.Commons.S3.Abstraction.csproj (hunk 1 above)
Comment (Reza-Noei) [reply Source/BSN.Commons.S3.Abstraction/BSN.Commons.S3.Abstraction.csproj]: I moved everything into BSN.Commons.

Looks much more better.

File: Source/BSN.Commons.S3.Abstraction/BucketPathBuilder.cs (hunk 2 above)
Comment (Reza-Noei) [reply Source/BSN.Commons.S3.Abstraction/BucketPathBuilder.cs]: I have removed this to make everything more clear.

File: Source/BSN.Commons.S3.Abstraction/BulkCredentialModel.cs (hunk 3 above)
Comment (Reza-Noei) [reply Source/BSN.Commons.S3.Abstraction/BulkCredentialModel.cs]: I've added comment for it please review it.

File: Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj (hunk 5 above)
Comment (Reza-Noei) [reply Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj]: I have argued the issues with you and we decided to accept the issue with .netframework.

Comment (Reza-Noei) [reply Source/BSN.Commons.S3.Net/BSN.Commons.S3.Net.csproj]: Now it's a part of the BSN.Commons.

**Review by soroshsabz — APPROVED** [review]

**Pull request merged**